## Prototype 2

In [2]:
import pandas as pd

#### Loading a csv file into a dataframe and displaying the first 5 rows

In [6]:
df = pd.read_csv("E:/SCHOOL/Portfolio/Real_Estate_Sales_2001-2021_GL.csv")
df.head(3)

C:\Users\JERRY\AppData\Local\Temp\ipykernel_4720\1289392730.py:1: DtypeWarning: Columns (8,9,10,11,12) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("E:/SCHOOL/Portfolio/Real_Estate_Sales_2001-2021_GL.csv")


Serial Number  List Year Date Recorded     Town          Address  \
0        2020348       2020    09/13/2021  Ansonia  230 WAKELEE AVE   
1          20002       2020    10/02/2020  Ashford  390 TURNPIKE RD   
2         210317       2021    07/05/2022     Avon  53 COTSWOLD WAY   

   Assessed Value  Sale Amount  Sales Ratio Property Type Residential Type  \
0        150500.0     325000.0       0.4630    Commercial              NaN   
1        253000.0     430000.0       0.5883   Residential    Single Family   
2        329730.0     805000.0       0.4096   Residential    Single Family   

  Non Use Code Assessor Remarks OPM remarks  \
0          NaN              NaN         NaN   
1          NaN              NaN         NaN   
2          NaN              NaN         NaN   

                             Location  
0                                 NaN  
1                                 NaN  
2  POINT (-72.846365959 41.781677018)

### Exploring the dataframe

#### Exploring the dataframe shape and the datatypes of the columns

In [12]:
df.shape

(1054159, 14)

In [10]:
df.dtypes

Serial Number         int64
List Year             int64
Date Recorded        object
Town                 object
Address              object
Assessed Value      float64
Sale Amount         float64
Sales Ratio         float64
Property Type        object
Residential Type     object
Non Use Code         object
Assessor Remarks     object
OPM remarks          object
Location             object
dtype: object

#### The dataset has 1054159 Rows and 14 Columns, with 2 columns of integer type, 3 columns of float type and 9 columns of object type 

#### Checking the number of missing values/Empty Spaces/NULL values in each column

In [16]:
missing_data = df.isnull()
for column in missing_data.columns.values.tolist():
    print (missing_data[column].value_counts())
    print("") 

Serial Number
False    1054159
Name: count, dtype: int64

List Year
False    1054159
Name: count, dtype: int64

Date Recorded
False    1054157
True           2
Name: count, dtype: int64

Town
False    1054159
Name: count, dtype: int64

Address
False    1054108
True          51
Name: count, dtype: int64

Assessed Value
False    1054159
Name: count, dtype: int64

Sale Amount
False    1054159
Name: count, dtype: int64

Sales Ratio
False    1054159
Name: count, dtype: int64

Property Type
False    671713
True     382446
Name: count, dtype: int64

Residential Type
False    660275
True     393884
Name: count, dtype: int64

Non Use Code
True     751917
False    302242
Name: count, dtype: int64

Assessor Remarks
True     892687
False    161472
Name: count, dtype: int64

OPM remarks
True     1042595
False      11564
Name: count, dtype: int64

Location
True     799516
False    254643
Name: count, dtype: int64



##### True represents NULL values/Empty spaces while 'False' represents valid values
#### The last four columns 'Non Use Code','Assessor Remarks','OPM Remarks' and 'Location' have higher percentage of NULL values,above 50% while 'Address' contains 51 and 'Date Recorded'  with 2 NULL values 

##### Non Use Code with 71.3%, Assessor Remarks: 84.7%, OPM Remarks: 98.9% and Location: 75.8%

###### The solution is to discard the three; 'Non Use Code', 'OPM Remarks' and 'Location' columns from the dataset

#### Removing the three columns: 'Non Use Code', 'OPM Remarks' and 'Location'

In [27]:
df.drop('Non Use Code', axis =1 , inplace=True)
df.drop('OPM remarks', axis =1 , inplace=True)
df.drop('Location', axis =1 , inplace=True)
df.head(3)

Serial Number  List Year Date Recorded     Town          Address  \
0        2020348       2020    09/13/2021  Ansonia  230 WAKELEE AVE   
1          20002       2020    10/02/2020  Ashford  390 TURNPIKE RD   
2         210317       2021    07/05/2022     Avon  53 COTSWOLD WAY   

   Assessed Value  Sale Amount  Sales Ratio Property Type Residential Type  \
0        150500.0     325000.0       0.4630    Commercial              NaN   
1        253000.0     430000.0       0.5883   Residential    Single Family   
2        329730.0     805000.0       0.4096   Residential    Single Family   

  Assessor Remarks  
0              NaN  
1              NaN  
2              NaN

#### 'Date Recorded' Column has two NULL values, showing the rows that corresponds to those two values

In [30]:
df[df['Date Recorded'].isnull()]

Serial Number  List Year Date Recorded    Town Address  \
234863          20280       2002           NaN  Orange     NaN   
236356              0       2002           NaN  Orange     NaN   

        Assessed Value  Sale Amount  Sales Ratio Property Type  \
234863             0.0          0.0          0.0           NaN   
236356             0.0          0.0          0.0           NaN   

       Residential Type Assessor Remarks  
234863              NaN              NaN  
236356              NaN              NaN

#### The two rows contains unuseful or redundant data hence we delete them

In [34]:
df = df.dropna(subset = ['Date Recorded'])
df[df['Date Recorded'].isnull()]

Empty DataFrame
Columns: [Serial Number, List Year, Date Recorded, Town, Address, Assessed Value, Sale Amount, Sales Ratio, Property Type, Residential Type, Assessor Remarks]
Index: []

##### Rows deleted

#### The address column now has 49 blank spaces / NULL values.Displaying them

In [39]:
df[df['Address'].isnull()]

Serial Number  List Year Date Recorded              Town Address  \
75944           39999       2003    02/02/2004        West Haven     NaN   
89004           49996       2004    05/17/2005            Lisbon     NaN   
101602          48886       2004    06/13/2005            Lisbon     NaN   
145969          10537       2001    02/05/2002          Hartford     NaN   
149638          10640       2001    12/19/2001        Bridgeport     NaN   
263008          30125       2003    11/10/2003       New Milford     NaN   
276118          39998       2003    08/12/2004            Lisbon     NaN   
284788          30100       2003    05/20/2004  North Stonington     NaN   
289386          39995       2003    02/02/2004        West Haven     NaN   
296156          39998       2003    02/20/2004        West Haven     NaN   
296748          40088       2004    11/01/2004            Groton     NaN   
297284          48889       2004    06/02/2005            Lisbon     NaN   
310485          40080       2004    11/24/2004        Brookfield     NaN   
313376          39997       2003    02/02/2004        West Haven     NaN   
315350          48888       2004    03/03/2005        Bridgeport     NaN   
319336          49999       2004    05/12/2005            Lisbon     NaN   
329404          48887       2004    06/30/2005            Lisbon     NaN   
330019          40070       2004    10/12/2004          Hartford     NaN   
331974          40059       2004    04/29/2005         Harwinton     NaN   
334893          48884       2004    05/13/2005            Lisbon     NaN   
338946          48885       2004    05/13/2005            Lisbon     NaN   
340556          49997       2004    05/27/2005            Lisbon     NaN   
341747          49994       2004    05/12/2005            Lisbon     NaN   
342559          48811       2004    06/22/2005            Lisbon     NaN   
344259          48810       2004    06/02/2005            Lisbon     NaN   
344747          49993       2004    05/23/2005            Lisbon     NaN   
344819          40318       2004    06/07/2005      East Hampton     NaN   
345481          48888       2004    06/08/2005            Lisbon     NaN   
345865          49998       2004    05/12/2005            Lisbon     NaN   
346288          41588       2004    09/20/2005           Meriden     NaN   
348686          49995       2004    05/12/2005            Lisbon     NaN   
365828          49999       2004    07/05/2005             Salem     NaN   
367924          49998       2004    07/05/2005             Salem     NaN   
368474          41230       2004    03/10/2005         Waterbury     NaN   
370597          48992       2004    05/12/2005             Salem     NaN   
372066          40198       2004    06/02/2005           Norfolk     NaN   
374438          49997       2004    07/08/2005             Salem     NaN   
391119          40285       2004    01/11/2005        Torrington     NaN   
396853          50862       2005    05/16/2006           Bristol     NaN   
462293          60058       2006    09/17/2007              Lyme     NaN   
466717          60474       2006    07/30/2007        Farmington     NaN   
471408          60054       2006    12/08/2006     New Fairfield     NaN   
471587          60032       2006    02/21/2007          Sterling     NaN   
471622          60159       2006    08/21/2007        Litchfield     NaN   
480138          60043       2006    07/19/2007           Pomfret     NaN   
499599          60237       2006    04/02/2007     South Windsor     NaN   
918926         170165       2017    12/08/2017        Manchester     NaN   
948271         172767       2017    01/12/2018           Shelton     NaN   
952087          17001       2017    10/02/2017       North Haven     NaN   

        Assessed Value  Sale Amount  Sales Ratio  Property Type  \
75944              0.0          0.0     0.000000            NaN   
89004              0.0          0.0     0.000000            NaN   
101602

##### The observation is,if the Assessed value and the Sales Amount are both 0,the data is redundant or unimportant or has no value hence the solution is to remove the rows.Showing the rows where 'Assessed Value' = 0 and 'Sale Amount' = 0

In [43]:
df[(df['Assessed Value'] == 0) & (df['Sale Amount'] == 0) ]

Serial Number  List Year Date Recorded        Town  \
45035           39004       2003    10/21/2003      Easton   
69868           30102       2003    06/09/2004   Thomaston   
73676           30130       2003    06/24/2004   Thomaston   
75944           39999       2003    02/02/2004  West Haven   
86518           49888       2004    07/12/2005  Bridgeport   
...               ...        ...           ...         ...   
449601          59999       2005    04/11/2006       Salem   
449811          50005       2005    10/04/2005      Putnam   
456476          59000       2005    12/01/2005      Orange   
463247          59010       2005    11/01/2005       Salem   
931886         170390       2017    03/19/2018   Greenwich   

                      Address  Assessed Value  Sale Amount  Sales Ratio  \
45035            15 FENSKY RD             0.0          0.0          0.0   
69868        76 LITCHFIELD ST             0.0          0.0          0.0   
73676          26-28 GROVE ST             0.0          0.0          0.0   
75944                     NaN             0.0          0.0          0.0   
86518         1650 RESERVOUIR             0.0          0.0          0.0   
...                       ...             ...          ...          ...   
449601       INDIANFIELD CAMP             0.0          0.0          0.0   
449811           224 GROVE ST             0.0          0.0          0.0   
456476        40 FARM HILL RD             0.0          0.0          0.0   
463247  211 OLD COLCHESTER RD             0.0          0.0          0.0   
931886        TWIN LAKES LANE             0.0          0.0          0.0   

       Property Type Residential Type     Assessor Remarks  
45035            NaN              NaN                  NaN  
69868            NaN              NaN                  NaN  
73676            NaN              NaN                  NaN  
75944            NaN              NaN                  NaN  
86518            NaN              NaN                  NaN  
...              ...              ...                  ...  
449601           NaN              NaN                  NaN  
449811           NaN              NaN                  NaN  
456476           NaN              NaN                  NaN  
463247           NaN              NaN                  NaN  
931886           NaN              NaN  multiple properties  

[285 rows x 11 columns]

#### Deleting the rows where both The Assessed Value and Sales Amount values is 0

In [46]:
condition = (df['Assessed Value'] == 0) & (df['Sale Amount'] == 0) 
df = df[~condition]
df[df['Address'].isnull()]

Serial Number  List Year Date Recorded              Town Address  \
145969          10537       2001    02/05/2002          Hartford     NaN   
149638          10640       2001    12/19/2001        Bridgeport     NaN   
263008          30125       2003    11/10/2003       New Milford     NaN   
284788          30100       2003    05/20/2004  North Stonington     NaN   
296748          40088       2004    11/01/2004            Groton     NaN   
310485          40080       2004    11/24/2004        Brookfield     NaN   
330019          40070       2004    10/12/2004          Hartford     NaN   
331974          40059       2004    04/29/2005         Harwinton     NaN   
344819          40318       2004    06/07/2005      East Hampton     NaN   
346288          41588       2004    09/20/2005           Meriden     NaN   
368474          41230       2004    03/10/2005         Waterbury     NaN   
372066          40198       2004    06/02/2005           Norfolk     NaN   
391119          40285       2004    01/11/2005        Torrington     NaN   
396853          50862       2005    05/16/2006           Bristol     NaN   
462293          60058       2006    09/17/2007              Lyme     NaN   
466717          60474       2006    07/30/2007        Farmington     NaN   
471408          60054       2006    12/08/2006     New Fairfield     NaN   
471587          60032       2006    02/21/2007          Sterling     NaN   
471622          60159       2006    08/21/2007        Litchfield     NaN   
480138          60043       2006    07/19/2007           Pomfret     NaN   
499599          60237       2006    04/02/2007     South Windsor     NaN   
918926         170165       2017    12/08/2017        Manchester     NaN   
948271         172767       2017    01/12/2018           Shelton     NaN   
952087          17001       2017    10/02/2017       North Haven     NaN   

        Assessed Value  Sale Amount  Sales Ratio  Property Type  \
145969             0.0     120000.0     0.000000            NaN   
149638       2106020.0      45000.0    46.800444            NaN   
263008         55090.0     400000.0     0.137725            NaN   
284788          7210.0     149000.0     0.048389            NaN   
296748             0.0    7060035.0     0.000000            NaN   
310485        147340.0    1015000.0     0.145163            NaN   
330019         54950.0     105000.0     0.523333            NaN   
331974             0.0     450000.0     0.000000            NaN   
344819             0.0     400000.0     0.000000            NaN   
346288             0.0        100.0     0.000000            NaN   
368474        234500.0     425000.0     0.551765            NaN   
372066         50200.0     350000.0     0.143429            NaN   
391119             0.0     155800.0     0.000000            NaN   
396853        104100.0     155000.0     0.671613            NaN   
462293             0.0       3656.0     0.000000            NaN   
466717             0.0     453369.0     0.000000  Single Family   
471408             0.0       3500.0     0.000000            NaN   
471587        181310.0     301500.0     0.601360            NaN   
471622         55190.0      60000.0     0.919833            NaN   
480138        445340.0     875000.0     0.508960            NaN   
499599         72340.0     225000.0     0.321511            NaN   
918926        129300.0     224000.0     0.577200     Two Family   
948271        227500.0     500000.0     0.455000          Condo   
952087        193130.0     242000.0     0.798000  Single Family   

       Residential Type Assessor Remarks  
145969              NaN              NaN  
149638              NaN              NaN  
263008              NaN              NaN  
284788              NaN              NaN  
296748              NaN              NaN  
310485              NaN              NaN  
330019              NaN              NaN  
331974              NaN              NaN  
344819              NaN              NaN  
3

#### Replacing Null values of the 'Address' Column

In [49]:
df['Address'].fillna('Unknown',inplace=True)

C:\Users\JERRY\AppData\Local\Temp\ipykernel_4720\3996658710.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Address'].fillna('Unknown',inplace=True)


In [51]:
missing_data = df['Address'].isnull()
print (missing_data.value_counts())
print("")

Address
False    1053872
Name: count, dtype: int64



### The 'Address' column now has no NULL values

#### Replacing the NULL values of the Property Type and Residential Type

In [63]:
df['Property Type'].fillna('Unknown',inplace=True)
df['Residential Type'].fillna('Non Residential',inplace=True)
df['Assessor Remarks'].fillna('No Comment',inplace=True)
df.head()

C:\Users\JERRY\AppData\Local\Temp\ipykernel_4720\3920804865.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Assessor Remarks'].fillna('No Comment',inplace=True)


Serial Number  List Year Date Recorded     Town                Address  \
0        2020348       2020    09/13/2021  Ansonia        230 WAKELEE AVE   
1          20002       2020    10/02/2020  Ashford        390 TURNPIKE RD   
2         210317       2021    07/05/2022     Avon        53 COTSWOLD WAY   
3         200212       2020    03/09/2021     Avon       5 CHESTNUT DRIVE   
4         200243       2020    04/13/2021     Avon  111 NORTHINGTON DRIVE   

   Assessed Value  Sale Amount  Sales Ratio Property Type Residential Type  \
0        150500.0     325000.0       0.4630    Commercial  Non Residential   
1        253000.0     430000.0       0.5883   Residential    Single Family   
2        329730.0     805000.0       0.4096   Residential    Single Family   
3        130400.0     179900.0       0.7248   Residential            Condo   
4        619290.0     890000.0       0.6958   Residential    Single Family   

  Assessor Remarks  
0       No Comment  
1       No Comment  
2       No Comment  
3       No Comment  
4       No Comment

#### Checking for all columns missing values

In [65]:
missing_data = df.isnull()
for column in missing_data.columns.values.tolist():
    print (missing_data[column].value_counts())
    print("")

Serial Number
False    1053872
Name: count, dtype: int64

List Year
False    1053872
Name: count, dtype: int64

Date Recorded
False    1053872
Name: count, dtype: int64

Town
False    1053872
Name: count, dtype: int64

Address
False    1053872
Name: count, dtype: int64

Assessed Value
False    1053872
Name: count, dtype: int64

Sale Amount
False    1053872
Name: count, dtype: int64

Sales Ratio
False    1053872
Name: count, dtype: int64

Property Type
False    1053872
Name: count, dtype: int64

Residential Type
False    1053872
Name: count, dtype: int64

Assessor Remarks
False    1053872
Name: count, dtype: int64



#### All the missing values and in all the columns are handled,no column contains the missing values

#### Handling the columns data types,first displaying each column's data type

In [70]:
df.dtypes

Serial Number         int64
List Year             int64
Date Recorded        object
Town                 object
Address              object
Assessed Value      float64
Sale Amount         float64
Sales Ratio         float64
Property Type        object
Residential Type     object
Assessor Remarks     object
dtype: object

##### Converting 'Date Recorded' column to type Date/Time,Town,Address,Property Type,Residential Type and Assessor Remarks to (str),Assessed Value and Sale Amount to (int)

In [76]:
df['Town'] = df['Town'].astype(str)
df['Address'] = df['Address'].astype(str)
df['Assessed Value'] = df['Assessed Value'].astype(int)
df['Sale Amount'] = df['Sale Amount'].astype(int)
df['Property Type'] = df['Property Type'].astype(str)
df['Residential Type'] = df['Residential Type'].astype(str)
df['Date Recorded'] = pd.to_datetime(df['Date Recorded'])
df['Assessor Remarks'] = df['Assessor Remarks'].astype(str)
print(df.dtypes)

Serial Number                int64
List Year                    int64
Date Recorded       datetime64[ns]
Town                        object
Address                     object
Assessed Value               int32
Sale Amount                  int32
Sales Ratio                float64
Property Type               object
Residential Type            object
Assessor Remarks            object
dtype: object


In [78]:
df.head()

Serial Number  List Year Date Recorded     Town                Address  \
0        2020348       2020    2021-09-13  Ansonia        230 WAKELEE AVE   
1          20002       2020    2020-10-02  Ashford        390 TURNPIKE RD   
2         210317       2021    2022-07-05     Avon        53 COTSWOLD WAY   
3         200212       2020    2021-03-09     Avon       5 CHESTNUT DRIVE   
4         200243       2020    2021-04-13     Avon  111 NORTHINGTON DRIVE   

   Assessed Value  Sale Amount  Sales Ratio Property Type Residential Type  \
0          150500       325000       0.4630    Commercial  Non Residential   
1          253000       430000       0.5883   Residential    Single Family   
2          329730       805000       0.4096   Residential    Single Family   
3          130400       179900       0.7248   Residential            Condo   
4          619290       890000       0.6958   Residential    Single Family   

  Assessor Remarks  
0       No Comment  
1       No Comment  
2       No Comment  
3       No Comment  
4       No Comment

#### 'Sales Ratio' column can be a calculated column hence for memory purposes,we delete it

In [81]:
df.drop('Sales Ratio', axis =1 , inplace=True)
df.head()

Serial Number  List Year Date Recorded     Town                Address  \
0        2020348       2020    2021-09-13  Ansonia        230 WAKELEE AVE   
1          20002       2020    2020-10-02  Ashford        390 TURNPIKE RD   
2         210317       2021    2022-07-05     Avon        53 COTSWOLD WAY   
3         200212       2020    2021-03-09     Avon       5 CHESTNUT DRIVE   
4         200243       2020    2021-04-13     Avon  111 NORTHINGTON DRIVE   

   Assessed Value  Sale Amount Property Type Residential Type Assessor Remarks  
0          150500       325000    Commercial  Non Residential       No Comment  
1          253000       430000   Residential    Single Family       No Comment  
2          329730       805000   Residential    Single Family       No Comment  
3          130400       179900   Residential            Condo       No Comment  
4          619290       890000   Residential    Single Family       No Comment

###### NB: BECAUSE THE DATASET IS TOO LARGE (1 053 872 ROWS) AND THIS IS A PERSONAL PROJECT FOR PORTFOLIO PURPOSES,WE WOULD LIKE TO DELETE THE ROWS WHERE WE REPLACED THE VALUES

In [83]:
condition = (df['Property Type'] == 'Unknown') 
df = df[~condition]
df[(df['Property Type'] == 'Unknown')] 

Empty DataFrame
Columns: [Serial Number, List Year, Date Recorded, Town, Address, Assessed Value, Sale Amount, Property Type, Residential Type, Assessor Remarks]
Index: []

In [85]:
condition = (df['Residential Type'] == 'Non Residential') 
df = df[~condition]
df[(df['Residential Type'] == 'Non Residential')]

Empty DataFrame
Columns: [Serial Number, List Year, Date Recorded, Town, Address, Assessed Value, Sale Amount, Property Type, Residential Type, Assessor Remarks]
Index: []

In [87]:
condition = (df['Assessor Remarks'] == 'No Comment')
df = df[~condition]
df[(df['Assessor Remarks'] == 'No Comment')]

Empty DataFrame
Columns: [Serial Number, List Year, Date Recorded, Town, Address, Assessed Value, Sale Amount, Property Type, Residential Type, Assessor Remarks]
Index: []

In [91]:
condition = (df['Address'] == 'Unknown')
df = df[~condition]
df[(df['Address'] == 'Unknown')]

Empty DataFrame
Columns: [Serial Number, List Year, Date Recorded, Town, Address, Assessed Value, Sale Amount, Property Type, Residential Type, Assessor Remarks]
Index: []

In [89]:
df.head()

Serial Number  List Year Date Recorded      Town               Address  \
15         200258       2020    2021-01-22  Branford  34 SUNRISE COVE CAMP   
25         201212       2020    2021-08-23   Bristol       8 JENNINGS TERR   
32        2000330       2020    2021-05-20  Cheshire      844 HIGHLAND AVE   
45         200742       2020    2021-03-29   Danbury      17 OHEHYAHTAH PL   
46         200886       2020    2021-05-03   Danbury      86 MERRIMAC ST B   

    Assessed Value  Sale Amount Property Type Residential Type  \
15          530500       700000   Residential    Single Family   
25           88060        91000   Residential    Single Family   
32          144220       204000   Residential    Single Family   
45          339100       550000   Residential    Single Family   
46          189800       240000   Residential            Condo   

                                     Assessor Remarks  
15  WATERFRONT/COTTAGE ONLY/LAND IS ASSOCIATION OWNED  
25                       FORECLOSURE SALE OUT OF BANK  
32                               DEFERRED MAINTENANCE  
45                                             H16095  
46                                           G16159-2

In [93]:
df.describe()

Serial Number      List Year                  Date Recorded  \
count   1.348100e+05  134810.000000                         134810   
mean    4.933618e+05    2014.731392  2015-12-30 17:06:01.842593280   
min     1.670000e+02    2006.000000            2006-10-02 00:00:00   
25%     9.127150e+04    2011.000000            2012-07-19 00:00:00   
50%     1.492635e+05    2015.000000            2016-04-18 00:00:00   
75%     1.901270e+05    2018.000000            2019-08-13 00:00:00   
max     2.111000e+08    2021.000000            2022-09-30 00:00:00   
std     4.559820e+06       4.098231                            NaN   

       Assessed Value   Sale Amount  
count    1.348100e+05  1.348100e+05  
mean     2.682252e+05  3.439672e+05  
min      0.000000e+00  0.000000e+00  
25%      1.023225e+05  1.060000e+05  
50%      1.526600e+05  1.970000e+05  
75%      2.333975e+05  3.437500e+05  
max      8.815100e+08  3.187900e+08  
std      2.658686e+06  1.661947e+06

In [97]:
df.shape

(134810, 10)

#### Checking if the Serial Number Column contains Duplicated Values

In [100]:
duplicated_data = df['Serial Number'].duplicated()
duplicated_data

15         False
25         False
32         False
45         False
46         False
           ...  
1054118     True
1054128     True
1054133     True
1054148     True
1054155     True
Name: Serial Number, Length: 134810, dtype: bool

##### The Serial Number contains Duplicated values

#### Printing the rows with duplicated values in a Serial Number Column

In [104]:
df[df['Serial Number'].duplicated(keep = False)]

Serial Number  List Year Date Recorded              Town  \
15              200258       2020    2021-01-22          Branford   
25              201212       2020    2021-08-23           Bristol   
32             2000330       2020    2021-05-20          Cheshire   
45              200742       2020    2021-03-29           Danbury   
46              200886       2020    2021-05-03           Danbury   
...                ...        ...           ...               ...   
1054118         190060       2019    2019-12-10            Wilton   
1054128          19035       2019    2020-02-05  North Stonington   
1054133         190478       2019    2020-02-14           Norwalk   
1054148        1900224       2019    2020-07-30           Wolcott   
1054155         190284       2019    2019-11-27         Waterbury   

                      Address  Assessed Value  Sale Amount  Property Type  \
15       34 SUNRISE COVE CAMP          530500       700000    Residential   
25            8 JENNINGS TERR           88060        91000    Residential   
32           844 HIGHLAND AVE          144220       204000    Residential   
45           17 OHEHYAHTAH PL          339100       550000    Residential   
46           86 MERRIMAC ST B          189800       240000    Residential   
...                       ...             ...          ...            ...   
1054118     72 WEST MEADOE RD          431060       385000  Single Family   
1054128          63 MYSTIC RD          192720       196300  Single Family   
1054133          38 HYATT AVE          283180       440000  Single Family   
1054148        12 OLD FARM RD          357680       490000  Single Family   
1054155       126 PERKINS AVE           68280        76000  Single Family   

        Residential Type                                   Assessor Remarks  
15         Single Family  WATERFRONT/COTTAGE ONLY/LAND IS ASSOCIATION OWNED  
25         Single Family                       FORECLOSURE SALE OUT OF BANK  
32         Single Family                               DEFERRED MAINTENANCE  
45         Single Family                                             H16095  
46                 Condo                                           G16159-2  
...                  ...                                                ...  
1054118    Single Family  BANK OWNED AND PRICED FOR IMMEDIATE SALE PER S...  
1054128    Single Family                                 FAMILY & FARM LAND  
1054133    Single Family                      FORECLOSURE BY SALE:  1-7-9-0  
1054148    Single Family                          MOTIVATED SELLER -DIVORCE  
1054155    Single Family                                       PRIVATE SALE  

[118151 rows x 10 columns]

#### Checking if the duplicated values duplicates the rows

In [107]:
dfD=df[df['Serial Number'].duplicated(keep = False)]
dfD=dfD.set_index('Serial Number')
dfD.duplicated()

Serial Number
200258     False
201212     False
2000330    False
200742     False
200886     False
           ...  
190060     False
19035      False
190478     False
1900224    False
190284     False
Length: 118151, dtype: bool

#### The Serial Number column has duplicated values but rows are not duplicated hence the Serial Number value cannot uniquely identify a row in our data frame¶

#### Since the Serial number does not uniquely identify each row,we drop the column and save index as a column

In [112]:
df.drop('Serial Number', axis =1 , inplace=True)
df.head()

List Year Date Recorded      Town               Address  Assessed Value  \
15       2020    2021-01-22  Branford  34 SUNRISE COVE CAMP          530500   
25       2020    2021-08-23   Bristol       8 JENNINGS TERR           88060   
32       2020    2021-05-20  Cheshire      844 HIGHLAND AVE          144220   
45       2020    2021-03-29   Danbury      17 OHEHYAHTAH PL          339100   
46       2020    2021-05-03   Danbury      86 MERRIMAC ST B          189800   

    Sale Amount Property Type Residential Type  \
15       700000   Residential    Single Family   
25        91000   Residential    Single Family   
32       204000   Residential    Single Family   
45       550000   Residential    Single Family   
46       240000   Residential            Condo   

                                     Assessor Remarks  
15  WATERFRONT/COTTAGE ONLY/LAND IS ASSOCIATION OWNED  
25                       FORECLOSURE SALE OUT OF BANK  
32                               DEFERRED MAINTENANCE  
45                                             H16095  
46                                           G16159-2

In [116]:
df.to_csv("E:/SCHOOL/Data Sets/archive/New folder/real_Estate_Final.csv", index = False)